Frameworks/UIKit/UIFontDescriptor.mm

```diff
@@ -29,233 +36,176 @@
-    fontDescriptor->_size = [UIFontDescriptor _getFallBackFontSize:size];
-    return [fontDescriptor autorelease];
-}
+@implementation UIFontDescriptor
 
-+ (UIFontDescriptor*)fontDescriptorWithDescriptor:(UIFontDescriptor*)descriptor size:(CGFloat)size {
-    UIFontDescriptor* fontDescriptor = [[UIFontDescriptor alloc] init];
-    fontDescriptor->_name = descriptor->_name;
-    fontDescriptor->_size = [UIFontDescriptor _getFallBackFontSize:size];
-    fontDescriptor->_traits = descriptor->_traits;
-    return [fontDescriptor autorelease];
-}
+BASE_CLASS_REQUIRED_IMPLS(UIFontDescriptor, UIFontDescriptorPrototype, CTFontDescriptorGetTypeID);
 
 /**
  @Status Caveat
-  @Notes         Force to use default font name and size.
+ @Notes  Force to use default font name and size.
 */
 + (UIFontDescriptor*)preferredFontDescriptorWithTextStyle:(NSString*)style {
     return [UIFontDescriptor fontDescriptorWithName:nil size:0.0];
 }
 
 /**
- This is a temp solution, we need to implement the fall back logic in 1512.
- Here just to check nil, and return default system font as fall back.
- */
-+ (NSString*)_getFallBackFontName:(NSString*)fontName {
-    NSString* ret = [NSString alloc];
-    if (fontName == nil || fontName.length == 0) {
-        ret = [ret initWithString:SystemFontName];
-    } else {
-        ret = [ret initWithString:fontName];
-    }
-    return [ret autorelease];
-}
-
-+ (float)_getFallBackFontSize:(float)size {
-    if (size <= 0) {
-        return kCTFontSystemFontSize;
-    } else {
-        return size;
-    }
-}
-
-/**
  @Status Interoperable
 */
-- (UIFontDescriptor*)fontDescriptorWithSymbolicTraits:(UIFontDescriptorSymbolicTraits)symbolicTraits {
-    UIFontDescriptor* result = [UIFontDescriptor fontDescriptorWithDescriptor:self size:self->_size];
-    result->_traits = symbolicTraits;
-    return result;
++ (UIFontDescriptor*)fontDescriptorWithFontAttributes:(NSDictionary*)attributes {
+    return [[[self alloc] initWithFontAttributes:attributes] autorelease];
 }
 
 /**
  @Status Stub
 */
-- (instancetype)initWithFontAttributes:(NSDictionary*)attributes {
++ (UIFontDescriptor*)fontDescriptorWithName:(NSString*)fontName matrix:(CGAffineTransform)matrix {
     UNIMPLEMENTED();
-    return nil;
-}
-
-- (NSString*)_getFontName {
-    return self->_name;
+    return StubReturn();
 }
 
 /**
  @Status Interoperable
 */
-- (UIFontDescriptorSymbolicTraits)symbolicTraits {
-    return self->_traits;
++ (UIFontDescriptor*)fontDescriptorWithName:(NSString*)fontName size:(CGFloat)size {
+    return [self fontDescriptorWithFontAttributes:@{
+        UIFontDescriptorNameAttribute : fontName,
+        UIFontDescriptorSizeAttribute : [NSNumber numberWithFloat:size]
+    }];
 }
 
 /**
  @Status Interoperable
 */
-- (void)dealloc {
-    _name = nil;
-    [super dealloc];
+- (UIFontDescriptor*)fontDescriptorByAddingAttributes:(NSDictionary*)attributes {
+    NSMutableDictionary* newAttributes = [[self fontAttributes] mutableCopy];
```

Comment:
oops

---
In reply to: [78862671](https://github.com/Microsoft/WinObjC/pull/1008#discussion_r78862671) [](ancestors = 78862671)